cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -3862,20 +3862,41 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             ctx.emit(Inst::xmm_unary_rm_r(sse_op, src, dst));
         }
 
-        Opcode::Fpromote => {
-            // We can't guarantee the RHS (if a load) is 128-bit aligned, so we
-            // must avoid merging a load here.
-            let src = RegMem::reg(put_input_in_reg(ctx, inputs[0]));
-            let dst = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
-            ctx.emit(Inst::xmm_unary_rm_r(SseOpcode::Cvtss2sd, src, dst));
+        Opcode::Fpromote | Opcode::FvpromoteLow => {
+            let output_ty = ty.unwrap();
+            if !output_ty.is_vector() {
+                // We can't guarantee the RHS (if a load) is 128-bit aligned, so we
+                // must avoid merging a load here.
+                let src = RegMem::reg(put_input_in_reg(ctx, inputs[0]));
+                let dst = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
+                ctx.emit(Inst::xmm_unary_rm_r(SseOpcode::Cvtss2sd, src, dst));
+            } else {
+                let src = RegMem::reg(put_input_in_reg(ctx, inputs[0]));
+                let dst = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
+                ctx.emit(Inst::xmm_unary_rm_r(
+                    SseOpcode::Cvtps2pd,
+                    RegMem::from(src),
+                    dst,
+                ));
```

Comment:
This allows `fpromote.f64x2` and `fpromote_low.f64x2` to compile to the same x64 instruction, which I don't think is intended. Perhaps assert that the vectors have `op == Opcode::FpromoteLow` and the scalars have `op == Opcode::Fpromote`. (This comment also applies below if you don't do the "fvdemote -> fdemote" renaming).

Reply:
I think we still need to think about this comment...